File: app/services/relation_pool_builder.py
"""Deep module: build ranked 近義 / 反義 / 語意相關 pools (runtime + ingest 同源)."""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import List, Optional, Set

from sqlalchemy.orm import Session

from app.repositories.word_relation_repo import chars_present_in_db
from app.services.relation_graph import fetch_relation_tuples
from app.services.syn_ant_ranking import (
    dedupe_rel_items,
    final_score,
    merge_relation_pools,
    morpheme_chars_from_synonyms,
    morpheme_chars_from_word_lists,
    normalize_relation_row,
    sort_ant_pool,
    sort_syn_pool,
)
from app.services.thesaurus_port import ThesaurusPort, default_thesaurus_port


def _load_static_pools(query: str, thesaurus: ThesaurusPort) -> tuple[List[str], List[str]]:
    return thesaurus.get_synonyms(query), thesaurus.get_antonyms(query)


def _resolve_morpheme_chars(
    query: str,
    static_syns: List[str],
    static_ants: List[str],
    thesaurus: ThesaurusPort,
) -> Set[str]:
    if len(query) < 2:
        return set()
    try:
        return morpheme_chars_from_synonyms(thesaurus.get_synonyms(query))
    except Exception:
        pass
    if static_syns:
        return morpheme_chars_from_word_lists(static_syns, static_ants)
    return set()


def _static_relation_pool(relation: str, words: List[str], present: Set[str]) -> List[dict]:
    return [
        {
            "char": w,
            "relation": relation,
            "source": "runtime_static",
            "score": None,
            "in_db": w in present,
            "jyutping": "",
            "code": "",
            "_sort": final_score(source="runtime_static", confidence=0.5, in_db=w in present),
        }
        for w in words
    ]


def _apply_in_db_membership(items: List[dict], present: Set[str]) -> List[dict]:
    out: List[dict] = []
    for item in items:
        ch = item.get("char") or ""
        in_db = ch in present
        patched = dict(item)
        patched["in_db"] = in_db
        patched["_sort"] = final_score(
            source=patched.get("source"),
            confidence=patched.get("score"),
            in_db=in_db,
        )
        out.append(patched)
    return out


def _fetch_db_relations(db: Session, query: str) -> List[dict]:
    q = query.strip()
    if not q:
        return []
    items: List[dict] = []
    for row in fetch_relation_tuples(db, q):
        item = normalize_relation_row(*row, query=q, db_char_set=set())
        if item:
            items.append(item)
    items = dedupe_rel_items(items)
    items.sort(key=lambda x: (x.get("_sort", 99), x.get("char") or ""))
    return items


def _collect_sorted_pool(
    *,
    query: str,
    relation: str,
    rel_items: List[dict],
    static_words: List[str],
    present: Set[str],
    morpheme_chars: Set[str],
) -> List[dict]:
    out: List[dict] = []
    seen: Set[str] = set()
    db_pool = [i for i in rel_items if i["relation"] == relation]
    static_pool = _static_relation_pool(relation, static_words, present)
    if relation == "syn":
        effective_morphemes = morpheme_chars if len(query) >= 2 else set()
        pool = sort_syn_pool(
            query,
            list(merge_relation_pools(db_pool, static_pool).values()),
            effective_morphemes,
        )
    elif relation == "ant":
        effective_morphemes = morpheme_chars if len(query) >= 2 else set()
        pool = sort_ant_pool(
            query,
            list(merge_relation_pools(db_pool, static_pool).values()),
            effective_morphemes,
        )
    else:
        pool = db_pool + static_pool
        pool.sort(key=lambda x: (x.get("_sort", 99), x.get("char") or ""))

    for item in pool:
        ch = item.get("char") or ""
        if not ch or ch in seen or ch == query:
            continue
        seen.add(ch)
        out.append(item)
    return out


@dataclass
class RelationPoolSnapshot:
    """Sorted pool items for one 字面 query."""

    query: str
    syn_pool: List[dict]
    ant_pool: List[dict]
    semantic_pool: List[dict]
    rel_syn: int = 0
    rel_ant: int = 0
    rel_sem: int = 0
    static_syn: int = 0
    static_ant: int = 0

    def chars(self, kind: str) -> List[str]:
        if kind == "syn":
            rows = self.syn_pool
        elif kind == "ant":
            rows = self.ant_pool
        else:
            return []
        return [r["char"] for r in rows if r.get("char")]


class RelationPoolBuilder:
    """Build syn/ant/語意相關 pools — same sources as 近反義模式 / 近義橋反義 ingest."""

    def __init__(
        self,
        db: Session,
        *,
        thesaurus: Optional[ThesaurusPort] = None,
        membership: Optional[Set[str]] = None,
    ) -> None:
        self._db = db
        self._thesaurus = thesaurus or default_thesaurus_port()
        self._membership = membership

    def build(self, query: str, *, include_static: bool = True) -> RelationPoolSnapshot:
        if not query or not re.search(r"[\u4e00-\u9fff]", query):
            return RelationPoolSnapshot(query=query or "", syn_pool=[], ant_pool=[], semantic_pool=[])

        q = query.strip()
        rel_items: List[dict] = []
        try:
            rel_items = _fetch_db_relations(self._db, q)
        except Exception:
            rel_items = []

        static_syns: List[str] = []
        static_ants: List[str] = []
        if include_static:
            static_syns, static_ants = _load_static_pools(q, self._thesaurus)
        morpheme_chars = _resolve_morpheme_chars(q, static_syns, static_ants, self._thesaurus)

        candidate_chars: Set[str] = set()
        for item in rel_items:
            ch = item.get("char")
            if ch:
                candidate_chars.add(ch)
        candidate_chars.update(static_syns)
        candidate_chars.update(static_ants)

        if self._membership is not None:
            present = self._membership
        else:
            present = chars_present_in_db(self._db, candidate_chars)
        rel_items = _apply_in_db_membership(rel_items, present)

        syn_pool = _collect_sorted_pool(
            query=q,
            relation="syn",
            rel_items=rel_items,
            static_words=static_syns,
            present=present,
            morpheme_chars=morpheme_chars,
        )
        ant_pool = _collect_sorted_pool(
            query=q,
            relation="ant",
            rel_items=rel_items,
            static_words=static_ants,
            present=present,
            morpheme_chars=morpheme_chars,
        )

        seen_main = {q} | {r["char"] for r in syn_pool} | {r["char"] for r in ant_pool}
        semantic_pool: List[dict] = []
        for item in rel_items:
            if item.get("relation") != "semantic_related":
                continue
            ch = item.get("char") or ""
            if ch and ch not in seen_main:
                seen_main.add(ch)
                semantic_pool.append(item)

        return RelationPoolSnapshot(
            query=q,
            syn_pool=syn_pool,
            ant_pool=ant_pool,
            semantic_pool=semantic_pool,
            rel_syn=sum(1 for i in rel_items if i["relation"] == "syn"),
            rel_ant=sum(1 for i in rel_items if i["relation"] == "ant"),
            rel_sem=sum(1 for i in rel_items if i["relation"] == "semantic_related"),
            static_syn=len(static_syns),
            static_ant=len(static_ants),
        )


__all__ = [
    "RelationPoolBuilder",
    "RelationPoolSnapshot",
]

File: client/src/db/relation-pool-builder.ts
/** 近反義池建構 — port of domain/relations/pool_builder.py (Phase C PR4). */
import type { Database } from './sqljs.ts';
import { queryRows } from './database-backend.ts';
import { getStaticAntonyms, getStaticSynonyms } from './thesaurus.ts';
import { getCuratedAntCompounds } from './compound.ts';
import { appendRuntimeDerivedAntPool } from './derived-ant.ts';
import {
  createRelationPoolSnapshot,
  type RelationKind,
  type RelationPoolItem,
  type RelationPoolSnapshot,
} from './relation-pool-snapshot.ts';
import {
  RUNTIME_DERIVED_ANT_SOURCES,
  finalScore,
  mergeRelationPools,
  morphemeCharsFromWordLists,
  sortAntPool,
  sortSynPool,
} from './relation-pool-ranking.ts';

const CJK_RE = /[\u4e00-\u9fff]/

export function poolLiteral(text: string): string | null {
  const t = (text ?? '')
    .trim()
    .replace(/[（(].*?[）)]/g, '')
    .replace(/\s+/g, '');
  if (!t || t.length > 12 || !CJK_RE.test(t) || /[0-9A-Za-z_]/.test(t)) {
    return null;
  }
  return t;
}
export function parseGroupCodes(raw: unknown): string[] {
  if (!raw) {
    return [];
  }
  if (Array.isArray(raw)) {
    return raw.map(String).filter(Boolean);
  }
  if (typeof raw === 'string') {
    try {
      const parsed = JSON.parse(raw);
      if (Array.isArray(parsed)) {
        return parsed.map(String).filter(Boolean);
      }
    } catch {
      return raw ? [raw] : [];
    }
  }
  return [];
}

export function filterStaticWords(words: string[]): string[] {
  const out: string[] = [];
  const seen = new Set<string>();
  for (const w of words) {
    const t = poolLiteral(w);
    if (t && !seen.has(t)) {
      seen.add(t);
      out.push(t);
    }
  }
  return out;
}
export async function charsPresentInDb(db: Database, chars: Iterable<string>): Promise<Set<string>> {
  const unique = [...new Set(chars)].filter(Boolean);
  if (!unique.length) {
    return new Set();
  }
  const present = new Set<string>();
  const chunk = 500;
  for (let i = 0; i < unique.length; i += chunk) {
    const part = unique.slice(i, i + chunk);
    const placeholders = part.map(() => '?').join(',');
    const rows = await queryRows(db, `SELECT DISTINCT char FROM words WHERE char IN (${placeholders})`, part);
    for (const row of rows) {
      const ch = String((row as { char?: string }).char ?? '');
      if (ch) {
        present.add(ch);
      }
    }
  }
  return present;
}

export async function loadDbCharSet(db: Database): Promise<Set<string>> {
  const rows = await queryRows(db, 'SELECT DISTINCT char FROM words');
  const out = new Set<string>();
  for (const row of rows) {
    const ch = String((row as { char?: string }).char ?? '');
    if (ch) {
      out.add(ch);
    }
  }
  return out;
}

const BIDIRECTIONAL_REL_ROWS_SQL = `
  SELECT wr.relation_type AS relation_type, w2.char AS rchar, wr.source AS source,
         wr.score AS score, w2.jyutping AS jyutping, w2.code AS code, wr.group_codes AS group_codes
  FROM words w1
  JOIN word_relations wr ON wr.word_id = w1.id
  JOIN words w2 ON w2.id = wr.related_id
  WHERE w1.char = ? AND wr.relation_type IN ('syn','ant','semantic_related')
  UNION ALL
  SELECT wr.relation_type, w1.char, wr.source, wr.score, w1.jyutping, w1.code, wr.group_codes
  FROM words w2
  JOIN word_relations wr ON wr.related_id = w2.id
  JOIN words w1 ON w1.id = wr.word_id
  WHERE w2.char = ? AND wr.relation_type IN ('syn','ant','semantic_related')
`;

export async function fetchDbRelations(db: Database, query: string): Promise<RelationPoolItem[]> {
  const q = query.trim();
  if (!q) {
    return [];
  }
  const rows = await queryRows(db, BIDIRECTIONAL_REL_ROWS_SQL, [q, q]);
  const items: RelationPoolItem[] = [];
  for (const row of rows) {
    const source = String(row.source ?? 'word_relations');
    if (RUNTIME_DERIVED_ANT_SOURCES.has(source)) {
      continue;
    }
    const rchar = poolLiteral(String(row.rchar ?? ''));
    if (!rchar || rchar === q) {
      continue;
    }
    const relation = String(row.relation_type ?? '') as RelationKind;
    if (!['syn', 'ant', 'semantic_related'].includes(relation)) {
      continue;
    }
    const groupCodes = parseGroupCodes(row.group_codes);
    items.push({
      char: rchar,
      relation,
      source,
      score: row.score == null ? null : Number(row.score),
      in_db: false,
      jyutping: String(row.jyutping ?? ''),
      code: String(row.code ?? ''),
      group_codes: groupCodes,
      _sort: finalScore(source, row.score == null ? null : Number(row.score), false),
    });
  }
  const best = new Map<string, RelationPoolItem>();
  for (const item of items) {
    const key = `${item.char}\t${item.relation}`;
    const prev = best.get(key);
    if (!prev || item._sort < prev._sort) {
      best.set(key, item);
    }
  }
  return [...best.values()];
}

export function staticRelationPool(
  relation: RelationKind,
  words: string[],
  present: Set<string>,
): RelationPoolItem[] {
  return filterStaticWords(words).map((w) => ({
    char: w,
    relation,
    source: 'runtime_static',
    score: null,
    in_db: present.has(w),
    jyutping: '',
    code: '',
    group_codes: [],
    _sort: finalScore('runtime_static', 0.5, present.has(w)),
  }));
}

export function applyInDbMembership(items: RelationPoolItem[], present: Set<string>): RelationPoolItem[] {
  return items.map((item) => {
    const inDb = present.has(item.char);
    return {
      ...item,
      in_db: inDb,
      _sort: finalScore(item.source, item.score, inDb),
    };
  });
}

export function collectSortedPool(
  query: string,
  relation: 'syn' | 'ant',
  relItems: RelationPoolItem[],
  staticWords: string[],
  present: Set<string>,
  morphemeChars: Set<string>,
): RelationPoolItem[] {
  const dbPool = relItems.filter((i) => i.relation === relation);
  const staticPool = staticRelationPool(relation, staticWords, present);
  const effectiveMorphemes = query.length >= 2 ? morphemeChars : new Set<string>();
  const merged =
    relation === 'syn'
      ? sortSynPool(query, [...mergeRelationPools(dbPool, staticPool).values()], effectiveMorphemes)
      : sortAntPool(query, [...mergeRelationPools(dbPool, staticPool).values()], effectiveMorphemes);

  const out: RelationPoolItem[] = [];
  const seen = new Set<string>();
  for (const item of merged) {
    const ch = item.char;
    if (!ch || ch === query || seen.has(ch)) {
      continue;
    }
    seen.add(ch);
    out.push(item);
  }
  return out;
}

export async function buildRelationPool(
  db: Database,
  query: string,
  options: { includeStatic?: boolean; includeDerivedAnt?: boolean } = {},
): Promise<RelationPoolSnapshot> {
  const includeStatic = options.includeStatic !== false;
  const includeDerivedAnt = options.includeDerivedAnt !== false;
  const q = query.trim();
  if (!q || !CJK_RE.test(q)) {
    return createRelationPoolSnapshot(q, [], [], []);
  }

  let relItems = await fetchDbRelations(db, q);
  let staticSyns: string[] = [];
  let staticAnts: string[] = [];
  if (includeStatic) {
    staticSyns = filterStaticWords(getStaticSynonyms(q));
    staticAnts = filterStaticWords(getStaticAntonyms(q));
    if (q.length === 1) {
      const extra: string[] = [];
      for (const compound of getCuratedAntCompounds()) {
        if (compound.length !== 2 || !compound.includes(q)) {
          continue;
        }
        const other = compound[0] === q ? compound[1]! : compound[0]!;
        if (other && other !== q) {
          extra.push(other);
        }
      }
      staticAnts = [...new Set([...staticAnts, ...extra])];
    }
  }
  const morphemeChars =
    q.length >= 2
      ? morphemeCharsFromWordLists(staticSyns, staticAnts)
      : new Set<string>();

  const candidateChars = new Set<string>();
  for (const item of relItems) {
    candidateChars.add(item.char);
  }
  for (const w of staticSyns) {
    candidateChars.add(w);
  }
  for (const w of staticAnts) {
    candidateChars.add(w);
  }

  const present = await charsPresentInDb(db, candidateChars);
  relItems = applyInDbMembership(relItems, present);

  const synPool = collectSortedPool(q, 'syn', relItems, staticSyns, present, morphemeChars);
  let antPool = collectSortedPool(q, 'ant', relItems, staticAnts, present, morphemeChars);

  if (includeDerivedAnt) {
    const membership = await loadDbCharSet(db);
    const headSyns = new Set(synPool.map((r) => r.char));
    const relAntRows = relItems
      .filter((i) => i.relation === 'ant')
      .map((i) => ({ char: i.char, source: i.source }));
    const effectiveMorphemes = q.length >= 2 ? morphemeChars : new Set<string>();
    antPool = sortAntPool(
      q,
      await appendRuntimeDerivedAntPool(
        q,
        antPool,
        db,
        membership,
        includeStatic,
        effectiveMorphemes,
        headSyns,
        relAntRows,
      ),
      effectiveMorphemes,
    ).filter((item) => item.char && item.char !== q);
  }

  const seenMain = new Set([q, ...synPool.map((r) => r.char), ...antPool.map((r) => r.char)]);
  const semanticPool = relItems.filter((item) => {
    if (item.relation !== 'semantic_related') {
      return false;
    }
    const ch = item.char;
    if (!ch || seenMain.has(ch)) {
      return false;
    }
    seenMain.add(ch);
    return true;
  });

  return createRelationPoolSnapshot(q, synPool, antPool, semanticPool);
}
